### PR TITLE
Let authorities vet identities and only in case of sufficient LoA

### PR DESCRIPTION
**Pull request by rjkip**

### CODE REVIEW

**Comment by rjkip** [comment]
:facepunch: @DRvanR

File: src/Surfnet/StepupMiddleware/CommandHandlingBundle/Tests/Identity/CommandHandler/IdentityCommandHandlerTest.php [hunk 1]
```diff
@@ -631,4 +634,202 @@ public function an_unverified_second_factor_can_be_revoked()
...
+                    $authorityId,
+                    $authorityInstitution,
+                    $authorityPhoneSfId,
+                    DateTime::now(),
+                    'regcode',
+                    'Charlie Parker',
+                    'e@mail.com',
+                    'en_GB'
+                ),
+                new SecondFactorVettedEvent(
```
Comment (DRvanR) [inline src/Surfnet/StepupMiddleware/CommandHandlingBundle/Tests/Identity/CommandHandler/IdentityCommandHandlerTest.php]: Seems simpler to use the bootstrap event?

Comment (rjkip) [inline src/Surfnet/StepupMiddleware/CommandHandlingBundle/Tests/Identity/CommandHandler/IdentityCommandHandlerTest.php]: The bootstrap event bootstraps a Yubikey, which has a higher LoA.

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 2]
```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
...
-            )) {
-                continue;
-            }
+    public function vetSecondFactor(
+        IdentityApi $identity,
+        SecondFactorId $secondFactorId,
+        $secondFactorIdentifier,
+        $registrationCode,
+        $documentNumber,
+        $identityVerified
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: Would be nice to wrap the last three params in a VO rather then keep passing em around. Also, I wonder if `identityVerified` should be passed down all the way to the SF, rather than do it here. Semantically it might be right to do as well: `If I have not verified the Identity, I may not vet this SecondFactor`

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 3]
```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
...
+    ) {
+        /** @var VettedSecondFactor|false $secondFactorWithHighestLoa */
+        $secondFactorWithHighestLoa = array_reduce(
+            $this->vettedSecondFactors->toArray(),
+            function (VettedSecondFactor $carry, VettedSecondFactor $item) {
+                return $carry->hasEqualOrHigherLoaComparedTo($item) ? $carry : $item;
+            },
+            $this->vettedSecondFactors->first()
+        );
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: What would you think about moving this into a custom `SecondFactorCollection` that extends `ArrayCollection` or something. Seems it might be a bit cleaner? Fairly complex code to read... just a thought.

Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: Also, would be easier to mask/abstract that this currently doesn't do anything, since there can be only one (Vetted)SecondFactor at this time....

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 4]
```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
...
 
-        throw new DomainException(
-            "Cannot vet second factor: no second factor with given registration code or registration window has closed."
+        $secondFactor->vet(
+            $registrationCode,
+            $secondFactorIdentifier,
+            $documentNumber,
+            $identityVerified,
+            $this->vettedSecondFactors,
+            $this->vettedSecondFactors
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: Can't seem to find the usage of these additional params?

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 5]
```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
...
+
+        if (!$secondFactor->wouldBeVettedBy(
+            $registrationCode,
+            $secondFactorIdentifier,
+            $documentNumber,
+            $identityVerified
+        )) {
+            throw new DomainException(
+                'Cannot vet second factor: registration code is incorrect or registration window has closed.'
+            );
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: Either the checks should be done here and the DomainException should be made precise, of this should be guards of invariants in the SecondFactor and the specific exceptions should be thrown there. Reading the previous sentence to me there is only one solution, kinda like with the email verification. Thoughts?

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 6]
```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
...
-                $secondFactorIdentifier,
-                $documentNumber,
-                $identityVerified
-            )) {
-                continue;
-            }
+    public function vetSecondFactor(
+        IdentityApi $identity,
+        SecondFactorId $secondFactorId,
+        $secondFactorIdentifier,
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: I would recommend renaming `$identity` to `$registrant` (indicative of the fact that that identity is the person standing at the desk requesting the vetting). Same with `$secondFactorId` and `$secondFactorType`, it should be clear that those are related to the registrant and not this identity :wink:

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 7]
```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
...
-            return;
+        $verifiedSecondFactorHasEqualOrLowerLoaComparedTo =
+            $identity->verifiedSecondFactorHasEqualOrLowerLoaComparedTo(
+                $secondFactorId,
+                $secondFactorWithHighestLoa
+            );
+
+        if (!$verifiedSecondFactorHasEqualOrLowerLoaComparedTo) {
+            throw new DomainException('Authority does not have the required LoA to vet the identity\'s second factor');
+        }
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: Wall of Text inc!

TL;DR: not sure how I feel about this all, but it's just a feeling. Have a different implementation below, how would you feel about something like that?

---

In all honesty, I'm not sure how I feel about these lines.... I've been hesitant about whether or not I should say something since it's just a feeling and it may well be that it's since I'm reading it on github, but these lines (and their consequences) are very hard to read/grok; amongst things due to the long variable and method names and the triple dispatch behind it. 
It kinda feels like that it could be simpler by just requesting the Verified Second Factor from the Identity and checking the LoA's of both... something like:

```php
$secondFactorToBeVetted = $identityToBeVetted->getVerifiedSecondFactor($sfid);
if (!$mySecondFactor->hasLoaHigherOrEqualTo($secondFactorToBeVetted) {
    throw new DomainException();
}
```
Which is somewhat analogous to what is happening in reality... 
The upside of something like that would also be that you get to call `$secondFactorToBeVetted->vet(/* params */)` albeit that might well be a bridge too far...

Also, `$identity->complyWithVettingOfSecondFactor` could also be `$identity->complyWithVetting()` as the Identity itself is vetted as well :wink:

---

With the renaming and SecondFactorCollection this would become something like:
```php
// when we have to check based on current LoA (future) this might become 
// $this->vettedSecondFactors->get('xxx');
$mySecondFactor = $this->vettedSecondFactors->getSecondFactorWithHighestLoa();
$registrantSecondFactor = $registrant->getVerifiedSecondFactor($registrantSecondFactorId);

if (!$mySecondFactor->hasLoaHigherThanOrEqualTo($registrantSecondFactor()) {
    throw new DomainException('naughty boy!');
}

$registrant->complyWithVetting(/* params */);
```

Thoughts?

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 2 above)
Comment (rjkip) [inline src/Surfnet/Stepup/Identity/Identity.php]: If I put this in a VO, I'd not make things difficult by not passing it to the second factor.

File: src/Surfnet/StepupMiddleware/CommandHandlingBundle/Tests/Identity/CommandHandler/IdentityCommandHandlerTest.php (hunk 1 above)
Comment (DRvanR) [inline src/Surfnet/StepupMiddleware/CommandHandlingBundle/Tests/Identity/CommandHandler/IdentityCommandHandlerTest.php]: check

Comment (DRvanR) [inline src/Surfnet/StepupMiddleware/CommandHandlingBundle/Tests/Identity/CommandHandler/IdentityCommandHandlerTest.php]: sucks when doing stuff good makes other things harder :tongue:

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 2 above)
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: It's not about difficulty, it's about whose responsibility it is... Even if it is part of the VO, should it be the responsibility of the second-factor-to-be-vetted to check whether or not the identity on which this method is invoked (authority) has actually verified the identity (registrant) that second-factor-to-be-vetted belongs to, hence the question (akin to the remarks about the `complyWithVetting` check/exception)

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 3 above)
Comment (rjkip) [inline src/Surfnet/Stepup/Identity/Identity.php]: Moving into a custom collection sounds fine. What do you mean by "masking/abstracting that it currently doesn't do anything"?

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 5 above)
Comment (rjkip) [inline src/Surfnet/Stepup/Identity/Identity.php]: The difference between e-mail verification and vetting is that e-mail verification may verify one of the potentially multiple second factors. In the case of vetting, the exact second factor is known beforehand, which eliminates the need to check from the outside whether the second factor would be vetted by. `VerifiedSecondFactor#vet()` checks the preconditions itself, throwing specific `DomainException`s. I posit there's actually no need for up-front checking of vettability and all these checks can be removed.

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 6 above)
Comment (rjkip) [inline src/Surfnet/Stepup/Identity/Identity.php]: I assume you mean `$secondFactorIdentifier` instead of `$secondFactorType`?

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 3 above)
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: There is currently exactly one vetted second factor, rendering the `array_reduce` a bit redundant, not much to reduce if there is just one element - this could be masked by using the custom collection that actually returns just the one element it holds directly rather than using the reduce. Not that that has to happen, but would be fine to do.

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 7 above)
Comment (rjkip) [inline src/Surfnet/Stepup/Identity/Identity.php]: I agree with fetching the verified second factor to remove the triple dispatch.
I disagree with calling `$sf->vet()`, let's not mess with child entities from the outside.
I disagree with renaming the method to `complyWithVetting` for consistency reasons: the caller is called `vetSecondFactor` and the event (important, hard to change) is called `SecondFactorVetted`.

Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: Fair enough, good arguments :+1:

File: src/Surfnet/Stepup/Identity/Identity.php (hunk 6 above)
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: yup, my bad

File: src/Surfnet/Stepup/Identity/Entity/SecondFactorCollection.php [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+{
+    /**
+     * @return SecondFactor|null
+     */
+    public function getSecondFactorWithHighestLoa()
+    {
+        return array_reduce(
+            $this->toArray(),
+            function (SecondFactor $carry, SecondFactor $item) {
+                return $carry->hasEqualOrHigherLoaComparedTo($item) ? $carry : $item;
```
Comment (rjkip) [inline src/Surfnet/Stepup/Identity/Entity/SecondFactorCollection.php]: I kept the reduction as the collection is unaware that it currently will contain 0..1 second factors.

**Comment by rjkip** [comment]
Better? @DRvanR :dancer:

File: src/Surfnet/Stepup/Identity/Entity/SecondFactorCollection.php (hunk 8 above)
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Entity/SecondFactorCollection.php]: fair enough

File: src/Surfnet/Stepup/Identity/Entity/SecondFactorCollection.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+            },
+            $this->first() ?: null
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Entity/SecondFactorCollection.php]: Won't this break if it is `null`? (see typehint and method invocation in anonymous function)

File: src/Surfnet/Stepup/Identity/Entity/VerifiedSecondFactor.php [hunk 10]
```diff
@@ -159,6 +144,16 @@ public function vet($registrationCode, $secondFactorIdentifier, $documentNumber,
         );
     }
 
+    public function hasEqualOrHigherLoaComparedTo(SecondFactor $comparable)
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Entity/VerifiedSecondFactor.php]: repeated three times in the exact same manner, smells like abstraction could be done?

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 11]
```diff
@@ -235,25 +235,64 @@ public function verifyEmail($verificationNonce)
...
+        $registrationCode,
+        $documentNumber,
+        $identityVerified
+    ) {
+        /** @var VettedSecondFactor|null $secondFactorWithHighestLoa */
+        $secondFactorWithHighestLoa = $this->vettedSecondFactors->getSecondFactorWithHighestLoa();
+        $registrantsSecondFactor = $registrant->getVerifiedSecondFactor($registrantsSecondFactorId);
+
+        if ($registrantsSecondFactor === null) {
+            throw new DomainException("Registrant second factor by given ID does not exist");
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: sprintf the ID in?

File: src/Surfnet/Stepup/Identity/Identity.php [hunk 12]
```diff
@@ -235,25 +235,64 @@ public function verifyEmail($verificationNonce)
...
+        }
+
+        if (!$secondFactorWithHighestLoa->hasEqualOrHigherLoaComparedTo($registrantsSecondFactor)) {
+            throw new DomainException('Authority does not have the required LoA to vet the identity\'s second factor');
```
Comment (DRvanR) [inline src/Surfnet/Stepup/Identity/Identity.php]: `s/identity's/registrants/`

File: src/Surfnet/StepupMiddleware/CommandHandlingBundle/Identity/CommandHandler/IdentityCommandHandler.php [hunk 13]
```diff
@@ -175,15 +175,21 @@ public function handleVerifyEmailCommand(VerifyEmailCommand $command)
...
     {
+        /** @var IdentityApi $authority */
+        $authority = $this->repository->load(new IdentityId($command->authorityId));
         /** @var IdentityApi $identity */
         $identity = $this->repository->load(new IdentityId($command->identityId));
-        $identity->vetSecondFactor(
-            $command->registrationCode,
+
+        $authority->vetSecondFactor(
+            $identity,
```
Comment (DRvanR) [inline src/Surfnet/StepupMiddleware/CommandHandlingBundle/Identity/CommandHandler/IdentityCommandHandler.php]: registrant?

**Comment by DRvanR** [comment]
Some minor details, looks good otherwise! :+1:
